### PR TITLE
Webapp service name detector

### PR DESCRIPTION
**Pull request by breedx-splk**
This is the start of a few contributions of `ResourceDetector` implementations from the Splunk distro. The first one coming up here is the `AppServerServiceNameProvider`. This will eventually detect when the `service.name` resource attribute is missing (and also `otel.service.name` not set in config) and will apply a series of heuristics for common application servers.

Because the current implementation is not exactly small, it's being contributed piecemeal to make reviews easier. The implementation here doesn't actually apply any of the heuristics yet....to see what they'll look like when done, you can look in the [splunk distribution impl](https://github.com/signalfx/splunk-otel-java/tree/main/custom/src/main/java/com/splunk/opentelemetry/servicename).

### CODE REVIEW

File: settings.gradle.kts [hunk 1]
```diff
@@ -50,6 +50,7 @@ include(":jmx-metrics")
 include(":maven-extension")
 include(":micrometer-meter-provider")
 include(":noop-api")
+include(":resource-providers")
```
[the remark this answers is not in the document]
Comment (breedx-splk) [reply settings.gradle.kts:53]: Derp, yeah, I intended to do that. Will do.

File: resource-providers/src/main/java/io/opentelemetry/resourceproviders/ServiceNameChecker.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * This class is a BeforeAgentListener that hooks into the agent lifecycle to determine if the
+ * otel.service.name has been set. It will log a warning to help the user better understand how to
+ * set their service name.
+ *
+ * <p>The serviceNameNotConfigured method is also used by some resource providers to determine if
+ * they should apply (eg. don't change the service name if already set).
+ */
+@AutoService(BeforeAgentListener.class)
+public class ServiceNameChecker implements BeforeAgentListener {
```
[the remark this answers is not in the document]
Comment (breedx-splk) [reply resource-providers/src/main/java/io/opentelemetry/resourceproviders/ServiceNameChecker.java:27]: Yeah, I also didn't know where best to put it. Felt weird here for sure. 

The class is kinda dual purpose -- it serves to log that warning at agent start but it also provides that static `serviceNameNotConfigured()` method that we're using from `shouldApply()`. I could be pretty easily convinced to remove the `@AutoService` and `BeforeAgentListener` stuff from the class, but then what do you think we should do with the remaining 3 line helper method? Just inline it?

[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply resource-providers/src/main/java/io/opentelemetry/resourceproviders/ServiceNameChecker.java:27]: > what do you think we should do with the remaining 3 line helper method? Just inline it?

yep, I'd just inline it. It does add a little copy-paste to the whole thing, but on the other hand the different components are clearly separated.

[the remark this answers is not in the document]
Comment (breedx-splk) [reply resource-providers/src/main/java/io/opentelemetry/resourceproviders/ServiceNameChecker.java:27]: Sounds good for now! Thanks for chatting it through.

**Review by mateuszrzeszutek — APPROVED** [review]

File: resource-providers/src/main/java/io/opentelemetry/resourceproviders/AppServer.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.resourceproviders;
```
[the remark this answers is not in the document]
Comment (breedx-splk) [reply resource-providers/src/main/java/io/opentelemetry/resourceproviders/AppServer.java:6]: So the `AppServerServiceNameProvider` will be the only real user-facing class (although the delegates could be too I suppose, if someone wanted to pull in only a subset of the supported app servers). For now, I think I should make the two interfaces package-private and that should help, right?

**Review by trask — APPROVED** [review]
> For now, I think I should make the two interfaces package-private and that should help, right?

looks good, we can revisit visibility issues later too, thx

**Pull request merged**